File: src/SIL.Machine.Shared/Translation/SegmentEditDistance.cs
using System.Collections.Generic;
using System.Linq;

namespace SIL.Machine.Translation
{
	internal class SegmentEditDistance : EditDistanceBase<string[], string>
	{
		private readonly WordEditDistance _wordEditDistance;

		public SegmentEditDistance()
		{
			_wordEditDistance = new WordEditDistance();
		}

		public double HitCost
		{
			get { return _wordEditDistance.HitCost; }
			set { _wordEditDistance.HitCost = value; }
		}

		public double SubstitutionCost
		{
			get { return _wordEditDistance.SubstitutionCost; }
			set { _wordEditDistance.SubstitutionCost = value; }
		}

		public double InsertionCost
		{
			get { return _wordEditDistance.InsertionCost; }
			set { _wordEditDistance.InsertionCost = value; }
		}

		public double DeletionCost
		{
			get { return _wordEditDistance.DeletionCost; }
			set { _wordEditDistance.DeletionCost = value; }
		}

		public double ComputePrefix(string[] x, string[] y, bool isLastItemComplete, bool usePrefixDelOp,
			out IEnumerable<EditOperation> wordOps, out IEnumerable<EditOperation> charOps)
		{
			double[,] distMatrix;
			double dist = Compute(x, y, isLastItemComplete, usePrefixDelOp, out distMatrix);

			charOps = null;
			int i = x.Length;
			int j = y.Length;
			var ops = new Stack<EditOperation>();
			while (i > 0 || j > 0)
			{
				EditOperation op;
				ProcessMatrixCell(x, y, distMatrix, usePrefixDelOp, j != y.Length || isLastItemComplete, i, j, out i,
					out j, out op);
				if (op != EditOperation.PrefixDelete)
					ops.Push(op);

				if (j + 1 == y.Length && !isLastItemComplete && op == EditOperation.Hit)
					_wordEditDistance.ComputePrefix(x[i], y[y.Length - 1], true, true, out charOps);
			}

			wordOps = ops;
			if (charOps == null)
				charOps = Enumerable.Empty<EditOperation>();

			return dist;
		}

		public void IncrComputePrefixFirstRow(List<double> scores, List<double> prevScores, string[] yIncr)
		{
			if (scores != prevScores)
			{
				scores.Clear();
				foreach (double score in prevScores)
					scores.Add(score);
			}

			int startPos = scores.Count;
			for (int jIncr = 0; jIncr < yIncr.Length; jIncr++)
			{
				int j = startPos + jIncr;
				if (j == 0)
					scores.Add(GetInsertionCost(yIncr[jIncr]));
				else
					scores.Add(scores[j - 1] + GetInsertionCost(yIncr[jIncr]));
			}
		}

		public IEnumerable<EditOperation> IncrComputePrefix(List<double> scores, List<double> prevScores, string xWord,
			string[] yIncr, bool isLastItemComplete)
		{
			var x = new[] {xWord};
			var y = new string[prevScores.Count - 1];
			for (int i = 0; i < yIncr.Length; i++)
				y[prevScores.Count - yIncr.Length - 1 + i] = yIncr[i];

			double[,] distMatrix = InitDistMatrix(x, y);

			for (int j = 0; j < prevScores.Count; j++)
				distMatrix[0, j] = prevScores[j];
			for (int j = 0; j < scores.Count; j++)
				distMatrix[1, j] = scores[j];

			while (scores.Count < prevScores.Count)
				scores.Add(0);

			int startPos = prevScores.Count - yIncr.Length;

			var ops = new List<EditOperation>();
			for (int jIncr = 0; jIncr < yIncr.Length; jIncr++)
			{
				int j = startPos + jIncr;
				int iPred, jPred;
				EditOperation op;
				double dist = ProcessMatrixCell(x, y, distMatrix, false, j != y.Length || isLastItemComplete,
					1, j, out iPred, out jPred, out op);
				scores[j] = dist;
				distMatrix[1, j] = dist;
				ops.Add(op);
			}

			return ops;
		}

		protected override int GetCount(string[] item)
		{
			return item.Length;
		}

		protected override string GetItem(string[] seq, int index)
		{
			return seq[index];
		}

		protected override double GetHitCost(string x, string y, bool isComplete)
		{
			return HitCost * y.Length;
		}

		protected override double GetSubstitutionCost(string x, string y, bool isComplete)
		{
			if (x == string.Empty)
				return (SubstitutionCost * 0.99) * y.Length;

			IEnumerable<EditOperation> ops;
			if (isComplete)
				_wordEditDistance.Compute(x, y, out ops);
			else
				_wordEditDistance.ComputePrefix(x, y, true, true, out ops);

			int hitCount, insCount, substCount, delCount;
			GetOpCounts(ops, out hitCount, out insCount, out substCount, out delCount);

			return (HitCost * hitCount) + (InsertionCost * insCount) + (SubstitutionCost * substCount)
				+ (DeletionCost * delCount);
		}

		private static void GetOpCounts(IEnumerable<EditOperation> ops, out int hitCount, out int insCount,
			out int substCount, out int delCount)
		{
			hitCount = 0;
			insCount = 0;
			substCount = 0;
			delCount = 0;
			foreach (EditOperation op in ops)
			{
				switch (op)
				{
					case EditOperation.Hit:
						hitCount++;
						break;
					case EditOperation.Insert:
						insCount++;
						break;
					case EditOperation.Substitute:
						substCount++;
						break;
					case EditOperation.Delete:
						delCount++;
						break;
				}
			}
		}

		protected override double GetDeletionCost(string x)
		{
			if (x == string.Empty)
				return DeletionCost;
			return DeletionCost * x.Length; 
		}

		protected override double GetInsertionCost(string y)
		{
			return InsertionCost * y.Length;
		}

		protected override bool IsHit(string x, string y, bool isComplete)
		{
			return x == y || (!isComplete && x.StartsWith(y));
		}
	}
}

File: src/SIL.Machine.TS/src/translation/segment-edit-distance.ts
import { EditDistance } from './edit-distance';
import { EditOperation } from './edit-operation';
import { WordEditDistance, WordEditDistanceResult } from './word-edit-distance';

export interface SegmentEditDistanceResult {
  cost: number;
  wordOps: EditOperation[];
  charOps: EditOperation[];
}

export class SegmentEditDistance extends EditDistance<string[], string> {
  private readonly wordEditDistance: WordEditDistance = new WordEditDistance();

  get hitCost(): number {
    return this.wordEditDistance.hitCost;
  }

  set hitCost(value: number) {
    this.wordEditDistance.hitCost = value;
  }

  get substitutionCost(): number {
    return this.wordEditDistance.substitutionCost;
  }

  set substitutionCost(value: number) {
    this.wordEditDistance.substitutionCost = value;
  }

  get insertionCost(): number {
    return this.wordEditDistance.insertionCost;
  }

  set insertionCost(value: number) {
    this.wordEditDistance.insertionCost = value;
  }

  get deletionCost(): number {
    return this.wordEditDistance.deletionCost;
  }

  set deletionCost(value: number) {
    this.wordEditDistance.deletionCost = value;
  }

  compute(x: string[], y: string[]): number {
    const matrixResult = this.computeDistMatrix(x, y, true, false);
    return matrixResult.cost;
  }

  computePrefix(
    x: string[],
    y: string[],
    isLastItemComplete: boolean,
    usePrefixDelOp: boolean
  ): SegmentEditDistanceResult {
    const matrixResult = this.computeDistMatrix(x, y, isLastItemComplete, usePrefixDelOp);

    let charOps: EditOperation[] = [];
    let i = x.length;
    let j = y.length;
    const wordOps: EditOperation[] = [];
    while (i > 0 || j > 0) {
      const cellResult = this.processDistMatrixCell(
        x,
        y,
        matrixResult.distMatrix,
        usePrefixDelOp,
        j !== y.length || isLastItemComplete,
        i,
        j
      );
      i = cellResult.iPred;
      j = cellResult.jPred;
      if (cellResult.op !== EditOperation.PrefixDelete) {
        wordOps.unshift(cellResult.op);
      }

      if (j + 1 === y.length && !isLastItemComplete && cellResult.op === EditOperation.Hit) {
        const wordResult = this.wordEditDistance.computePrefix(x[i], y[y.length - 1], true, true);
        charOps = wordResult.ops;
      }
    }

    return { cost: matrixResult.cost, wordOps, charOps };
  }

  incrComputePrefixFirstRow(scores: number[], prevScores: number[], yIncr: string[]): void {
    if (scores !== prevScores) {
      scores.length = 0;
      scores.push(...prevScores);
    }

    const startPos = scores.length;
    for (let jIncr = 0; jIncr < yIncr.length; jIncr++) {
      const j = startPos + jIncr;
      if (j === 0) {
        scores.push(this.getInsertionCost(yIncr[jIncr]));
      } else {
        scores.push(scores[j - 1] + this.getInsertionCost(yIncr[jIncr]));
      }
    }
  }

  incrComputePrefix(
    scores: number[],
    prevScores: number[],
    xWord: string,
    yIncr: string[],
    isLastItemComplete: boolean
  ): EditOperation[] {
    const x: string[] = [xWord];
    const y = new Array<string>(prevScores.length - 1);
    for (let i = 0; i < yIncr.length; i++) {
      y[prevScores.length - yIncr.length - 1 + i] = yIncr[i];
    }

    const distMatrix = this.initDistMatrix(x, y);

    for (let j = 0; j < prevScores.length; j++) {
      distMatrix[0][j] = prevScores[j];
    }
    for (let j = 0; j < scores.length; j++) {
      distMatrix[1][j] = scores[j];
    }

    while (scores.length < prevScores.length) {
      scores.push(0);
    }

    const startPos = prevScores.length - yIncr.length;

    const ops: EditOperation[] = [];
    for (let jIncr = 0; jIncr < yIncr.length; jIncr++) {
      const j = startPos + jIncr;
      const cellResult = this.processDistMatrixCell(
        x,
        y,
        distMatrix,
        false,
        j !== y.length || isLastItemComplete,
        1,
        j
      );
      scores[j] = cellResult.cost;
      distMatrix[1][j] = cellResult.cost;
      ops.push(cellResult.op);
    }

    return ops;
  }

  protected getCount(seq: string[]): number {
    return seq.length;
  }

  protected getItem(seq: string[], index: number): string {
    return seq[index];
  }

  protected getHitCost(x: string, y: string, isComplete: boolean): number {
    return this.hitCost * y.length;
  }

  protected getSubstitutionCost(x: string, y: string, isComplete: boolean): number {
    if (x === '') {
      return this.substitutionCost * 0.99 * y.length;
    }

    let result: WordEditDistanceResult;
    if (isComplete) {
      result = this.wordEditDistance.compute(x, y);
    } else {
      result = this.wordEditDistance.computePrefix(x, y, true, true);
    }

    const opCounts = this.getOpCounts(result.ops);

    return (
      this.hitCost * opCounts.hitCount +
      this.insertionCost * opCounts.insCount +
      this.substitutionCost * opCounts.substCount +
      this.deletionCost * opCounts.delCount
    );
  }

  protected getDeletionCost(x: string): number {
    if (x === '') {
      return this.deletionCost;
    }
    return this.deletionCost * x.length;
  }

  protected getInsertionCost(y: string): number {
    return this.insertionCost * y.length;
  }

  protected isHit(x: string, y: string, isComplete: boolean): boolean {
    return x === y || (!isComplete && x.startsWith(y));
  }

  private getOpCounts(
    ops: EditOperation[]
  ): { hitCount: number; insCount: number; substCount: number; delCount: number } {
    let hitCount = 0;
    let insCount = 0;
    let substCount = 0;
    let delCount = 0;
    for (const op of ops) {
      switch (op) {
        case EditOperation.Hit:
          hitCount++;
          break;
        case EditOperation.Insert:
          insCount++;
          break;
        case EditOperation.Substitute:
          substCount++;
          break;
        case EditOperation.Delete:
          delCount++;
          break;
      }
    }
    return { hitCount, insCount, substCount, delCount };
  }
}
